aten/src/ATen/core/dispatch/DispatchTable.h

```diff
@@ -205,40 +203,36 @@ class DispatchTable final {
     bool is_valid_;
 
     TensorTypeId get_dispatch_key(const Stack* stack, const std::string& operator_name) const {
-      const IValue& first_tensor_arg = torch::jit::peek(
-        *stack,
-        0,
-        reverse_index_of_first_tensor_arg_
-      );
-      // TODO: This will need to get adjusted for multiple dispatch
-      if (C10_UNLIKELY(first_tensor_arg_is_tensor_list_)) {
-        auto tensor_list = first_tensor_arg.toTensorListRef();
-        if (tensor_list.size() == 0) {
-          throw std::runtime_error("Tried to dispatch operator " + operator_name + " based on an empty tensor list. When the first tensor argument of an operator is a tensor list, then it must not be empty.");
+
+      TensorTypeSet ts;
+      for (const auto& ivalue : torch::jit::last(*stack, num_args_)) {
+        if (ivalue.isTensor()) {
+          ts = ts | ivalue.toTensor().type_set();
```

Comment:
This does a refcount bump. In my dispatch benchmarks, that was deadly for perf, which is why I the original code used `unsafeToTensorImpl()` to avoid the refcount bump. You should probably do that here too.

Reply:
I originally wrote `toTensorRef` here and was shocked that this didn't exist

Reply:
`toTensorRef` can't exist because `IValue` stores `TensorImpl` and not `Tensor`, so if you want a `const Tensor&`, you have to create a `Tensor` object first. You can only get `toTensorImplRef`, which is what `unsafeToTensorImpl` is doing.